README.md
sipgate.io
==========

This README documents the sipgate.io functionality. A demo page can be found [here](https://demo.sipgate.io).

Requirements
------------

### Usage with simquadrat

* [x] [Order a simquadrat SIM](https://www.simquadrat.de)
* [x] [Book the sipgate.io feature](https://www.simquadrat.de/feature-store/sipgate.io)
* [x] [Enter an URL for incoming/outgoing calls in the dashboard](https://www.simquadrat.de/dashboard)

### Usage with sipgate basic

* [x] [Create a free sipgate basic account](https://www.sipgate.de/go)
* [x] [Book the sipgate.io feature](https://www.sipgate.de/go/feature-store/sipgate.io)
* [x] [Enter an URL for incoming/outgoing calls in the dashboard](https://www.sipgate.de/go/dashboard)

### Usage with sipgate team (closed beta)

* [x] [Request access to our beta program](http://goo.gl/forms/8TS8kQj6kx)
* [x] [After receiving the confirmation mail enter an URL for incoming/outgoing calls in sipgate team settings](https://secure.live.sipgate.de/settings/sipgateio)


The POST request
================

sipgate.io sends POST requests with an `application/x-www-form-urlencoded` payload. Depending on the type of request it contains the following parameters:

### New call

Parameter | Description
--------- | -----------
from      | The calling number (e.g. `"492111234567"` or `"anonymous"`)
to        | The called number (e.g. `"4915791234567"`)
direction | The direction of the call (either `"in"` or `"out"`)
event     | "newCall"
callId    | A unique alphanumeric identifier to match events to specific calls

You can simulate this POST request and test your server with a cURL command:

```sh
curl -X POST --data "from=492111234567&to=4915791234567&direction=in&event=newCall&callId=123456" http://localhost:3000
```

### Call hangup

Parameter | Description
--------- | -----------
event     | "hangup"
callId    | Same as in newCall-Event for a specific call

You can simulate this POST request and test your server with a cURL command:

```sh
curl -X POST --data "event=hangup&callId=123456" http://localhost:3000
```



The XML response
============
After sending the POST request sipgate.io will accept an XML response to determine what to do. Make sure to set ```application/xml``` in the ```Content-Type``` header of your response.

sipgate.io currently supports the following responses for incoming and outgoing calls:

Action            | Description
----------------- | -----------
[Dial](#dial)     | Send call to voicemail or external number
[Play](#play)     | Play a sound file
[Reject](#reject) | Reject call or pretend to be busy
[Hangup](#hangup) | Hang up the call

Additional to actions, the response can specify urls which shall be called by sipgate.io on certain call-events. These urls are specified in form of xml-attributes in the response-tag. The only event implemented so far is hangup.

Url                   | Description
--------------------- | -----------
[onHangup](#onhangup) | Will receive a POST-request as soon as the call is ended. Either by hangup or cancel. The response to that request is discarded.

Dial
----

Redirect the call and alter your caller id ([call charges apply](https://www.simquadrat.de/tarife/mobile)).

Attribute | Possible values                                              | Default value
--------- | ------------------------------------------------------------ | -------------
callerId  | Number in [E.164](http://en.wikipedia.org/wiki/E.164) format | Account settings
anonymous | true, false                                                  | Account / phone settings

Possible targets for the dial command:

Target    | Description
--------- | -----------
Number    | Send call to an external number (has to be in [E.164](http://en.wikipedia.org/wiki/E.164) format)
Voicemail | Send call to [voicemail](https://www.simquadrat.de/feature-store/voicemail) (feature has to be booked)


**Example 1: Redirect call**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Dial>
		<Number>4915799912345</Number>
	</Dial>
</Response>
```

**Example 2: Send call to voicemail**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Dial>
		<Voicemail />
	</Dial>
</Response>
```

**Example 3: Suppress phone number**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Dial anonymous="true">
		<Number>4915799912345</Number>
	</Dial>
</Response>
```

**Example 4: Set custom caller id for outgoing call**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Dial callerId="492111234567">
		<!-- Originally dialed number, extracted from POST request -->
		<Number>4915799912345</Number>
	</Dial>
</Response>
```

Play
----

Play a given sound file.

Target    | Description
--------- | -----------
Url       | Play a sound file from a given URL


**Example 1: Play a sound file**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Play>
		<Url>http://example.com/example.wav</Url>
	</Play>
</Response>
```

**Please note:** Currently the sound file needs to be a mono 16bit PCM WAV file with a sampling rate of 8kHz. You can use conversion tools like the open source audio editor [Audacity](http://audacity.sourceforge.net/) to convert any sound file to the correct format.

Linux users might want to use ```mpg123``` to convert the file:
```sh
mpg123 --rate 8000 --mono -w output.wav input.mp3
```


Reject
------

Pretend to be busy or block unwanted calls.

Attribute | Possible values | Default value
--------- | --------------- | -------------
reason    | rejected, busy  | rejected


**Example 1: Reject call**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Reject />
</Response>
```

**Example 2: Reject call signaling busy**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Reject reason="busy" />
</Response>
```



Hangup
------

Hang up calls

**Example 1: Hang up call**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
	<Hangup />
</Response>
```

onHangup
------------

**Example 1: Request notification for call hangup**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response onHangup="http://localhost:3000/hangup">
</Response>
```


More to come
------------
Stay tuned...



Server Examples
===============

To get you started we maintain server examples for:

* [Node.js](https://github.com/sipgate/sipgate.io/tree/master/examples/nodejs)
* [PHP](https://github.com/sipgate/sipgate.io/tree/master/examples/php)

There are also examples in:

* [C++](https://github.com/sipgate/sipgate.io/tree/master/examples/c++)
* [Clojure](https://github.com/sipgate/sipgate.io/tree/master/examples/clojure)
* [Common Lisp](https://github.com/sipgate/sipgate.io/tree/master/examples/commonlisp)
* [Go](https://github.com/sipgate/sipgate.io/tree/master/examples/go)
* [Java](https://github.com/sipgate/sipgate.io/tree/master/examples/java)
* [Perl](https://github.com/sipgate/sipgate.io/tree/master/examples/perl)
* [Prolog](https://github.com/sipgate/sipgate.io/tree/master/examples/prolog)
* [Python](https://github.com/sipgate/sipgate.io/tree/master/examples/python)
* [Ruby](https://github.com/sipgate/sipgate.io/tree/master/examples/ruby)
* [Scala](https://github.com/sipgate/sipgate.io/tree/master/examples/scala)



Troubleshooting
===============

sipgate.io Log
--------------

You can enable logging for debugging purposes from your dashboard. You will find each request and the corresponding response in the logging table.

How do I inspect network traffic?
---------------------------------

You can use ```ngrep``` to inspect the incoming requests on your side:
```sh
sudo ngrep -dany -Wbyline port 3000
```



A word about security
=====================

HTTP vs. HTTPS
--------------

We strongly encourage you to use a HTTPS server. Although we support plain HTTP connections we do not recommend pushing sensitive call details over unencrypted connections. By default sipgate.io does not accept [self-signed certificates](http://stackoverflow.com/a/10176685), but you can allow them in your dashboard. Furthermore, you can add the public key of your certificate to protect the connection against man-in-the-middle attacks. Your certificate is validated by our server.

Authentication
--------------

sipgate.io supports HTTP Basic Authentication. You can include your username and password within the URL (e.g. `https://username:password@example.com:8080`).

Emergency calls
===============

sipgate.io does not process emergency calls. Emergency calls are immediately put through to emergency services.


Help us make it better
======================

Please tell us how we can improve sipgate.io. If you have a specific feature request, found a bug or would like to add an example, please use [GitHub Issues](https://github.com/sipgate/sipgate.io/issues) or fork these docs and send a [pull request](https://github.com/sipgate/sipgate.io/pulls) with your improvements.


CHANGELOG.md
sipgate.io Changelog
====================

1.6.0 Noah (2015-05-08)
-----------------------
***Features***
* sipgate.io available in basic
* sipgate.io available in team (beta)
* POST request for end of call

1.5.0 Emil (2014-11-27)
-----------------------
***Features***
* Set a custom caller id

1.4.0 Frida (2014-11-20)
-----------------------
***Features***
* POST request for outgoing calls

1.3.0 Phil (2014-11-18)
-----------------------
***Features***
* Play WAV file from a given URL
* ```<Play>``` ```<Url>``` example XML in README.md
* Hang up calls
* ```<Hangup>``` example XML in README.md
* Update examples

***Resolved issues***
* Client doesn't support compression, despite it being advertised in Accept-Encoding ([#11](https://github.com/sipgate/sipgate.io/issues/11))

1.2.0 Levi (2014-10-22)
-----------------------
***Features***
* Send call to voicemail or external number
* ```<Dial>``` ```<Number>``` example XML in README.md
* ```<Dial>``` ```<Voicemail>``` example XML in README.md
* Update examples

***Resolved issues***
* XML response not understood by sipgate.io's server ([#6](https://github.com/sipgate/sipgate.io/issues/6))
* Content-Type: text/xml support ([#7](https://github.com/sipgate/sipgate.io/issues/7))
* Chunked encoded response is not processed ([#8](https://github.com/sipgate/sipgate.io/issues/8))
* ```from``` and ```to``` not always correct ([#9](https://github.com/sipgate/sipgate.io/issues/9))

1.1.0 Nico (2014-10-14)
-----------------------
***Features***
* Reject calls with ```Reject```
* ```Reject``` example XML in README.md
* Update examples
* Add this cool CHANGELOG.md
* C++ example from  [Papierkorb](https://github.com/Papierkorb)

1.0.0 Brienne (2014-10-06)
--------------------------
***Features***
* First release
* POST request with ```from``` and ```to```
* Perl example from [fuglu](https://github.com/fuglu)
* Ruby example from [fuglu](https://github.com/fuglu)
* Python example from [scgitsip](https://github.com/scgitsip)
* PHP example from [tpxtron](https://github.com/tpxtron)
* Java example from [scgitsip](https://github.com/scgitsip)
* Go example from [vileda](https://github.com/vileada)
* Common Lisp example from [olivermg](https://github.com/olivermg)
* Clojure example from [olivermg](https://github.com/olivermg)
* Prolog example from [kr1schan](https://github.com/kr1schan)
* Scala example from [fnordian](https://github.com/fnordian)


examples/java/dial.java
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.StringWriter;
import java.net.InetSocketAddress;
import java.util.ArrayList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerConfigurationException;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;

import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import com.sun.net.httpserver.Headers;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpServer;

public class dial
{
	public static void main(String[] args) throws Exception
	{
		HttpServer httpServer = HttpServer.create(new InetSocketAddress(3000),0);
		httpServer.createContext("/", new AwesomeRejectHandler());
		httpServer.setExecutor(null);
		httpServer.start();

		System.out.println("Server running....");
		System.out.println("Press CTRL+C to stop the server");
	}

	static class AwesomeRejectHandler implements HttpHandler
	{
		public void handle(HttpExchange httpExchange) throws IOException
		{
			InputStream inputStream = httpExchange.getRequestBody();
			int charValue;
			ArrayList<Character> chars = new ArrayList<Character>();

			while ((charValue = inputStream.read()) != -1)
			{
				chars.add((char) charValue);
			}

			StringBuilder stringBuilder = new StringBuilder(chars.size());

			for (Character character : chars)
			{
				stringBuilder.append(character);
			}
			// Print from, to and direction
			System.err.println("INFO: DATA: " + stringBuilder.toString());

			String XMLResponse = "";
			XMLBuilder xmlBuilder;
			try
			{
				xmlBuilder = new XMLBuilder();
				if (stringBuilder.toString().contains("from=4915799912345"))
				{
					XMLResponse = xmlBuilder.dial("4915791234567");
				}
				else if (stringBuilder.toString().contains("from=491579987654"))
				{
					XMLResponse = xmlBuilder.send2voicemail();
				}
				else
				{
					XMLResponse = xmlBuilder.reject("busy");
				}
			}
			catch (ParserConfigurationException e)
			{
				e.printStackTrace();
			}

			// Print XML
			System.err.println("INFO: XML-Response: " + XMLResponse);

			Headers headers = httpExchange.getResponseHeaders();
			headers.set("Content-Type", "application/xml");
			headers.set("Accept-Encoding", "gzip");
			httpExchange.sendResponseHeaders(200, XMLResponse.length());
			OutputStream outputStream = httpExchange.getResponseBody();
			outputStream.write(XMLResponse.getBytes());
			outputStream.close();
		}
	}

	static class XMLBuilder
	{
		private Document document;
		private Element root;

		public XMLBuilder() throws ParserConfigurationException
		{
			DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
			DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();

			Document document = documentBuilder.newDocument();
			document.setXmlStandalone(true);
			//<Response> is the root tag 
			this.root = document.createElement("Response");
			document.appendChild(this.root);
			this.document = document;
		}

		private String dial(String number)
		{
			Element xmlNodeDial = document.createElement("Dial");
			root.appendChild(xmlNodeDial);
			Element xmlNodeNumber = document.createElement("Number");
			xmlNodeDial.appendChild(xmlNodeNumber);
			xmlNodeNumber.setTextContent(number);
			return build();
		}

		private String send2voicemail()
		{
			Element xmlNodeDial = document.createElement("Dial");
			root.appendChild(xmlNodeDial);
			Element xmlNodeVoicemail = document.createElement("Voicemail");
			xmlNodeDial.appendChild(xmlNodeVoicemail);
			return build();
		}
		
		private String reject(String reason)
		{
			Element xmlNodeDial = document.createElement("Reject");
			root.appendChild(xmlNodeDial);
			Attr xmlAttribute = document.createAttribute("reason");
			xmlAttribute.setValue(reason);
			xmlNodeDial.setAttributeNode(xmlAttribute);
			return build();
		}

		private String build()
		{
			String XMLResponse = "";

			try
			{
				// Write the document into XMl format and return it
				TransformerFactory transformerFactory = TransformerFactory.newInstance();
				Transformer transformer = transformerFactory.newTransformer();
				DOMSource domSource = new DOMSource(document);
				StringWriter stringWriter = new StringWriter();
				StreamResult streamResult = new StreamResult(stringWriter);
				transformer.transform(domSource, streamResult);
				XMLResponse = stringWriter.toString();
			}
			catch (TransformerException tranformerException)
			{
				tranformerException.printStackTrace();
			}
			return XMLResponse;
		}
	}
}
